### DISCUSSION TITLE
Is push RTP possible only on request?

### DISCUSSION THREAD

**Discussion by wolverin-a** [bluenviron/mediamtx#3230] [Questions]
### Question

Hello.
I'm trying to send a video through NAT to the server. BUT I need RTP packets to start being forwarded only by command (for example, PLAY)
Are there any ways to do this using mediamax using the RTSP protocol?

As I understand it now, mediamtx does not send any RTSP commands for this and the end of the dialogue looks like this, after which RTP starts to be sent immediately

```
- my application uses ffmpeg libraries
RECORD rtsp://remote_host:port/stream_url RTSP/1.0
Range: npt=0.000-
CSeq: 4
User-Agent: Lavf59.25.100
Session: 545498661

- mediamtx
RTSP/1.0 200 OK
CSeq: 4
Server: gortsplib
Session: 545498661
```

**Comment by aler9** [comment]
Hello, MediaMTX has been conceived for broadcasting, therefore no interactions between readers and the publisher of each stream are allowed, otherwise a single reader would influence all others. No RTSP commands can be forwarded from readers to the publisher.

However, if you're interested into forwarding RTP packets only when there are actually connected readers, you can do this by triggering some kind of event through `runOnRead`, like a HTTP request:

```yml
paths:
  mypath:
    runOnRead: curl http://remote_host:8080/start-fowarding-rtp-packets
```

You can then improve your application in order to listen for this HTTP request and behave accordingly.

**Reply by wolverin-a** [reply]
I can't do that because the remote_host is behind NAT.
Forwarding of all commands is not required, only PLAY is needed.

**Reply by aler9** [reply]
i did a little reasoning about your issue, in my opinion it can be solved with a pure RTSP solution by supporting back channels (#941). Unfortunately it's not ready yet, furthermore it would allow a SINGLE client to send data or commands back to the source, not any client.

If the remote_host is behind a NAT, my advice is to setup a [MQTT broker](https://mosquitto.org/) on the server in which you installed MediaMTX, connect your custom client to the broker, make it listen for a specific topic and send messages to the topic through mosquitto_pub:

```yml
paths:
  mypath:
    runOnRead: mosquitto_pub -t /commands -m "start-fowarding-rtp-packets"
```

**Reply by wolverin-a** [reply]
Thank you, I have already looked at the option with the MQTT broker, but in the option [WebRTC](https://github.com/meetecho/janus-gateway), which I can place on the "camera", perhaps this will even solve [my problem](https://github.com/bluenviron/mediamtx/issues/3308), since it is not necessary to forward RTP through a proxy server.

But the delay in the start of the video live stream is also important, and I will have to redo the clients.

[a comment by an automated account is not shown]
